pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionPartiallyOrderedByOperator.java

```diff
@@ -0,0 +1,303 @@
+  private final List<ExpressionContext> _alreadySorted;
+  private final List<ExpressionContext> _toSort;
+
+  private final TransformOperator _transformOperator;
+  private final List<OrderByExpressionContext> _orderByExpressions;
+  private final TransformResultMetadata[] _expressionsMetadata;
+  private final int _numRowsToKeep;
+  private final PartiallySortedListBuilder _sorter;
+
+  private int _numDocsScanned = 0;
+  private long _numEntriesScannedPostFilter = 0;
+  private boolean _used = false;
+  private Comparator<Object[]> _comparator;
+
+  public SelectionPartiallyOrderedByOperator(IndexSegment indexSegment, QueryContext queryContext,
+      List<ExpressionContext> expressions, TransformOperator transformOperator, int sortedExpr) {
+    Preconditions.checkArgument(sortedExpr > 0,
+        "This operator should not be used when sorting expressions doesn't start with sorted expressions");
+    _indexSegment = indexSegment;
+    _expressions = expressions;
+    _transformOperator = transformOperator;
+
+    _orderByExpressions = queryContext.getOrderByExpressions();
+    assert _orderByExpressions != null;
+    int numOrderByExpressions = _orderByExpressions.size();
+
+    _alreadySorted = expressions.subList(0, sortedExpr);
+    _toSort = expressions.subList(sortedExpr, numOrderByExpressions);
+
+    Preconditions.checkArgument(!_toSort.isEmpty(),
+        "This operator should not be used when all sorting expressions are sorted. "
+            + "Use %s instead", SelectionOrderByOperator.class.getName());
+
+    _expressionsMetadata = new TransformResultMetadata[_expressions.size()];
+    for (int i = 0; i < _expressionsMetadata.length; i++) {
+      ExpressionContext expression = _expressions.get(i);
+      _expressionsMetadata[i] = _transformOperator.getResultMetadata(expression);
+    }
+
+    _numRowsToKeep = queryContext.getOffset() + queryContext.getLimit();
+
+    int expectedSize = Math.min(_numRowsToKeep * 2, SelectionOperatorUtils.MAX_ROW_HOLDER_INITIAL_CAPACITY);
+    Comparator<Object[]> sortedComparator =
+        SelectionOrderByOperator.getComparator(_orderByExpressions, _expressionsMetadata, true, 0, sortedExpr);
+    Comparator<Object[]> unsortedComparator =
+        SelectionOrderByOperator.getComparator(_orderByExpressions, _expressionsMetadata, true, sortedExpr,
+            numOrderByExpressions);
+    _sorter = new PartiallySortedListBuilder(expectedSize, sortedComparator, unsortedComparator);
+    _comparator = SelectionOrderByOperator.getComparator(_orderByExpressions, _expressionsMetadata, false);
+  }
+
+  @Override
+  public List<Operator> getChildOperators() {
+    return Collections.singletonList(_transformOperator);
+  }
+
+  @Override
+  protected IntermediateResultsBlock getNextBlock() {
+    if (!_used) {
+      _used = true;
+      fetch();
+
+      DataSchema dataSchema = createDataSchema();
+
+      return new IntermediateResultsBlock.ListSelection(dataSchema, _sorter.build(), _comparator);
```

Comment:
Does it mean we have to build the sorter for every query per segment?

Reply:
We create one `SelectionPartiallyOrderedByOperator` per segment, right? So yeah, we build one of this `sorter`s for every segment.

Remember that `_sorter.build()` does not generate memory, it just returns the list it used internally and it shouldn't be used again.